### ISSUE TITLE
_pydstool_jl_hack - where is this?

### ISSUE THREAD

**Issue by jacobsca** [JuliaDiffEq/PyDSTool.jl#13]
I'm trying to use PyDSTool in DifferentialEquations...but it won't load. I get this long red error message that says 

ERROR: InitError: PyError (PyImport_ImportModule

The Python package _pydstool_jl_hack could not be found by pyimport. Usually this means
that you did not install _pydstool_jl_hack in the Python version being used by PyCall.

Any clues how to fix this?? I done some web crawling but cannot seemingly find a helpful answer...

**Comment by tkf** [comment]
@ChrisRackauckas No release AFAICT:
https://github.com/robclewley/pydstool/releases
https://pypi.org/project/PyDSTool/#history
https://anaconda.org/conda-forge/pydstool

Actually this "should not" happen.  But I realized that it can happen if pydstool or scipy is not installed.  I wasn't handling such case properly in `__init__` of PyDSTool.jl.

@jacobsca Maybe you are not using Conda.jl?  What is the output of the following commands?:

```julia
versioninfo()
using Pkg
Pkg.status()
using PyCall  # if it errors, try Pkg.add("PyCall") first
@show PyCall.conda
@show PyCall.pyprogramname
```

If `PyCall.conda` is `false` or `PyCall.pyprogramname` points to Python 3, you can try

```julia
using Pkg
ENV["CONDA_JL_VERSION"] = "2"
ENV["PYTHON"] = ""
Pkg.build("PyCall")  # let PyCall.jl use Conda.jl
exit()
```

to let Conda.jl download miniconda2 and then install pydstool in it.

**Comment by ChrisRackauckas** [comment]
Radau and Dopri needs a special install for PyDSTool that we don't do by default. These need to be compiled separately and something.

**Comment by HenriLaurie** [comment]
I think this is still unresolved. In my course at the moment we can't use PyDSTool because of both Python 2.x and Python 3.x being installed on the system. PyCall seems to be defaulting to Python 3.x and then using PyDSTool fails.

And often but not always it asks for _pdstool_jl_hack to be installed.

It isn't clear to me what we should do ... I am reluctant to play with ENV() because I worry it will contaminate other uses of PyCall.

**Comment by tkf** [comment]
I think using Python 2 is the only option unless you want to use pydstool master.  You need to use `ENV` for it since it is the official way to configure Python used for PyCall https://github.com/JuliaPy/PyCall.jl#specifying-the-python-version (i.e., `ENV["PYTHON"] = "python2"; using Pkg; Pkg.build("PyCall")`). As you worried, this will "contaminate" the PyCall setup because there is no way to use Python 2 without switching to it entirely.

**Comment by tkf** [comment]
> As for `Bifurcations.jl`, it still lacks the computation of Floquet compared to PyDSTool if I am not mistaken.

That's correct (sadly...).  Also, IIRC, PyDSTool (or more like AUTO) calculates the Jacobian for the bifurcation analysis/continuation of the limit cycle in a clever way (and uses its intermediate result for Floquet or something like that).  What I implemented there was super dumb (just use ForwardDiff).

**Comment by rveltz** [comment]
Yes they use a periodic schur decomposition. 

Yesterday, I was trying `Bifurcations.jl`, there is so much in the `test` that is not mentioned in the `docs`, it is a pity. A proper README.md would bring more traffic as well. Considering the amount of work and quality of this package, you should consider adding basic things like `Hopf_to_LC`... in the docs.

> What I implemented there was super dumb (just use ForwardDiff).

Do you mean that you have a way to get the Floquet coeffs? Even if not very accurate, you should add a quick and dirty solution, it can always be improved in the future.

**Comment by tkf** [comment]
> there is so much in the `test` that is not mentioned in the `docs`

Yeah, that's true...  The reason why I've been shy about this project was that I have hard time understanding some outputs.  Maybe there are critical bugs somewhere or I just miss-set some algorithm parameters or my intuition is wrong...  (In particular, https://github.com/tkf/Bifurcations.jl/pull/8 was the one behaving strangely. Though I need to clean it up and explain the oddity before asking for a help.)

Maybe I should just document everything and let people find the bugs for me :stuck_out_tongue:

> Do you mean that you have a way to get the Floquet coeffs?

I meant the computation of Jacobian is slow because I just simply hand it over to ForwardDiff.
